guide/src/migration.md

```diff
@@ -5,13 +5,91 @@ For a detailed list of all changes, see [CHANGELOG.md](https://github.com/PyO3/p
 
 ## from 0.11.* to 0.12
 
+### `PyErr` has been reworked
+
+In PyO3 `0.12` the `PyErr` type has been re-implemented to be significantly more compatible with
+the standard Rust error handling ecosystem. Specificially `PyErr` now implements
+`Error + Send + Sync`, which are the standard traits used for error types.
+
+While this has necessitated the removal of a number of APIs, the resulting `PyErr` type should now
+be much more easier to work with. The following sections list the changes in detail and how to
+migrate to the new APIs.
+
+#### `PyErr::new` and `PyErr::from_type` now require `Send + Sync` for their argument
+
+For most uses no change will be needed. If you are trying to construct `PyErr` from a value that is
+not `Send + Sync`, you will need to first create the Python object and then use
+`PyErr::from_instance`.
+
+Similarly, any types which implemented `PyErrArguments` will now need to be `Send + Sync`.
+
+#### `PyErr`'s contents are now private
+
+Making the `PyErr` internals private makes it possible for PyO3 to implement optimizations safely
+without needing to worry about breaking downstream code.
+
+If you needed to access the contents of the `PyErr`, you can use the new methods `PyErr::ptype()`,
+`PyErr::pvalue()` and `PyErr::ptraceback()` to safely do so.
+
+#### `PyErrValue` and `PyErr::from_value` have been removed
+
+As these were part the internals of `PyErr` which have been reworked, these APIs no longer exist.
+
+If you used this API, it is recommended to use `PyException::new_err` (see the next section on
+Exception types).
+
+#### `Into<PyResult<T>>` for `PyErr` has been removed
+
+This implementation was not idiomatic Rust. Just construct the `Result::Err` variant directly.
+
+Before:
+```rust,ignore
+let result: PyResult<()> = PyErr::new::<TypeError, _>("error message").into();
+```
+
+After (also using the new reworked exception types; see the following section):
+```rust
+# use pyo3::{PyErr, PyResult, exceptions::PyTypeError};
+let result: PyResult<()> = Err(PyTypeError::new_err("error message"));
+```
+
+### Exception types have been reworked
+
+Previously exception types were zero-sized marker types purely used to construct `PyErr`. In PyO3
+0.12, these types have been replaced with real definitions similar to `PyAny`, `PyDict` etc. This
```

Comment:
I feel ` real definitions similar to ...` a bit difficult.
What is important is we can use them like `&PyAny`, so I think `these type now has real definitions and is usable like PyAny, PyDict etc.` would be better.